File: 02.06.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @return Boolean
     */
    function isPalindrome($head) {
        $fast = $head;
        $slow = $head;
        while ($fast) {
            $fast = $fast->next->next;
            $slow = $slow->next;
        }
        $cur = null;
        while ($slow) {
            $next = $slow->next;
            $slow->next = $cur;
            $cur = $slow;
            $slow = $next;
        }
        while ($cur) {
            if ($cur->val !== $head->val) {
                return false;
            }
            $cur = $cur->next;
            $head = $head->next;
        }
        return true;
    }
}

File: go/linkedList/02.06.go
package main
//
//func isPalindrome(head *ListNode) bool {
//	fast := head.Next
//	slow := head
//
//	for fast.Next != nil && fast != nil {
//		fast = fast.Next.Next
//		slow = slow.Next
//	}
//	right := slow.Next
//	slow.Next = nil
//	left := head
//
//	var cur *ListNode
//	for right != nil {
//		next := right.Next
//		right.Next = cur
//		cur = right
//		right = next
//	}
//	for cur != nil {
//		if cur.Val != left.Val {
//			return false
//		}
//		cur = cur.Next
//		left = left.Next
//	}
//	return true
//}
//
//type ListNode struct {
//	Val  int
//	Next *ListNode
//}
